pkg/volume/downwardapi/downwardapi.go

```diff
@@ -0,0 +1,329 @@
+
+// collectData collects requested metadata in data map.
+// Map's key is the requested name of file to dump
+// Map's value is the (sorted) content of the field to be dumped in the file.
+func (m *downwardAPIVolume) collectData(data *map[string]string) error {
+	errlist := []error{}
+	for fieldReference, fileName := range m.fieldReferenceFileNames {
+		if values, err := fieldpath.ExtractFieldPathAsString(m.pod, fieldReference); err != nil {
+			glog.Error(err)
+			errlist = append(errlist, err)
+		} else {
+			(*data)[fileName] = sortLines(values)
+		}
+	}
+	return utilErrors.NewAggregate(errlist)
+}
+
+// isDataChanged iterate over all the entries to check wether at least one
+// file needs to be updated.
+func (m *downwardAPIVolume) isDataChanged(data *map[string]string) bool {
+	for fileName, values := range *data {
+		if isFileToGenerate(path.Join(m.GetPath(), fileName), values) {
+			return true
+		}
+	}
+	return false
+}
+
+// isFileToGenerate compares actual file with the new values. If
+// different (or the file does not exist) return true
+func isFileToGenerate(fileName, values string) bool {
+	if _, err := os.Lstat(fileName); os.IsNotExist(err) {
+		return true
+	}
+	return readFile(fileName) != values
+}
+
+const currentDir = ".current"
+const currentTmpDir = ".current_tmp"
+
+// writeData writes requested metadata in specified files.
+//
+// The file visible in this volume are symlinks to files in the '.current'
+// directory. Actual files are stored in an hidden timestamped directory which is
+// symlinked to by '.current'. The timestamped directory and '.current' symlink
+// are created in the plugin root dir.  This scheme allows the files to be
+// atomically updated by changing the target of the '.current' symlink.  When new
+// data is available:
+//
+// 1.  A new timestamped dir is created by writeDataInTimestampDir
+// 2.  Symlinks for new files are created if needed
+// 3.  The previous timestamped directory is detected reading the '.current' symlink
+// 4.  In case no symlink exists then it's created and func returns (first execution)
+// 5.  In case symlink exists a new temporary symlink is created .current_tmp
+// 6.  .current_tmp is renamed to .current
+// 7.  The previous timestamped directory is removed
+func (m *downwardAPIVolume) writeData(data *map[string]string) error {
+	var timestampDir string
+	var err error
+	timestampDir, err = m.writeDataInTimestampDir(data)
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+	// update symbolic links for relative paths
+	if err = m.updateSymlinksToCurrentDir(); err != nil {
+		os.RemoveAll(timestampDir)
+		glog.Error(err)
+		return err
+	}
+
+	_, timestampDirBaseName := filepath.Split(timestampDir)
+	var oldTimestampDirectory string
+	oldTimestampDirectory, err = os.Readlink(path.Join(m.GetPath(), currentDir))
+	if err != nil {
+		if os.IsNotExist(err) { // no link to currentDir so creates it
+			err = os.Symlink(timestampDirBaseName, path.Join(m.GetPath(), currentDir))
```

Comment:
Why special case this?  Would it be simpler to just let this fall through to the tmplink + rename path?  You need to read the old link to remove it, but why have two paths that produce the same result?